### PR TITLE
Crypto security

### PR DESCRIPTION
**Pull request by jaredbroad**
Implementation of a crypto security type by @jameschch

### CODE REVIEW

File: Common/Securities/Crypto/Crypto.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        /// <param name="symbolProperties">The symbol properties for this security</param>
+        public Crypto(SecurityExchangeHours exchangeHours, Cash quoteCurrency, SubscriptionDataConfig config, SymbolProperties symbolProperties)
+            : base(config,
+                quoteCurrency,
+                symbolProperties,
+                new CryptoExchange(exchangeHours),
+                new ForexCache(),
+                new SecurityPortfolioModel(),
+                new ImmediateFillModel(),
+                new InteractiveBrokersFeeModel(),
```
Comment (jaredbroad) [inline Common/Securities/Crypto/Crypto.cs]: Perhaps constant fee model (0)? I thought most exchanges were spread based?

File: Common/Brokerages/DefaultBrokerageModel.cs [hunk 2]
```diff
@@ -155,6 +156,9 @@ public decimal GetLeverage(Security security)
                 case SecurityType.Cfd:
                     return 50m;
 
+                case SecurityType.Crypto:
+                    return 3m;
```
Comment (jaredbroad) [inline Common/Brokerages/DefaultBrokerageModel.cs]: Perhaps default leverage should be 1 as most exchanges/brokerages don't support leverage yet? Later we can add custom leverage by brokerage model.

File: Common/Orders/OrderJsonConverter.cs [hunk 3]
```diff
@@ -105,8 +106,19 @@ public static Order CreateOrderFromJObject(JObject jObject)
...
+            //does data have market?
+            var suppliedMarket = jObject.SelectTokens("Symbol.ID.Market");
+            if (suppliedMarket.Any())
+            {
+                market = suppliedMarket.Single().Value<string>();
+            }
+            else
+            {
+                //no data, use default                
+                new DefaultBrokerageModel().DefaultMarkets.TryGetValue(securityType, out market);
```
Comment (jaredbroad) [inline Common/Orders/OrderJsonConverter.cs]: This is nice.

File: Common/Securities/Cash.cs [hunk 4]
```diff
@@ -179,7 +180,7 @@ public Security EnsureCurrencyDataFeed(SecurityManager securities, SubscriptionM
             var currencyPairs = Currencies.CurrencyPairs.Select(x =>
             {
                 // allow XAU or XAG to be used as quote currencies, but pairs including them are CFDs
-                var securityType = Symbol.StartsWith("X") ? SecurityType.Cfd : SecurityType.Forex;
+                var securityType = Symbol.StartsWith("XAU") ? SecurityType.Cfd : SecurityType.Forex;
```
Comment (jaredbroad) [inline Common/Securities/Cash.cs]: I think this drops XAG as quote currency, might need second If statement.

File: Common/Brokerages/DefaultBrokerageModel.cs (hunk 2 above)
Comment (jameschch) [reply Common/Brokerages/DefaultBrokerageModel.cs]: It's only really the US regulated exchanges that cannot offer leverage. But I see your point, we should not treat gdax behaviour as the default for crypto.

File: Common/Securities/Crypto/Crypto.cs (hunk 1 above)
Comment (jameschch) [reply Common/Securities/Crypto/Crypto.cs]: The gdax model is fee-free for makers after rebates. The other exchanges I have in mind, OKCoin, Bitfinex and Kraken have a maker/taker split at around 0.01%/0.02%. The Chinese exchanges used to offer fee free until the PBOC "visits" earlier this year. It makes sense to provide a constant fee model with the expectation that it needs overriding as anyone writing an algorithm for crypto that does not specify a valid fee model will get uselessly biased results.

File: Common/Securities/Cash.cs (hunk 4 above)
Comment (jameschch) [reply Common/Securities/Cash.cs]: Fixed

**Comment by jameschch** [comment]
Changes are here:

[https://github.com/QuantConnect/Lean/pull/1133](https://github.com/QuantConnect/Lean/pull/1133)

**Pull request merged**